src/ints/mouse.cpp

```diff
@@ -18,1549 +18,932 @@
+        DEBUG_QUEUE("ActivateIRQ, in %s", __FUNCTION__);
+        // If no timer in progress, handle the event now
+        PIC_ActivateIRQ(12);
+    }
+}
+
+void MouseQueue::AggregateEventsDOS(MouseEvent &event)
 {
-	if ((px != 0) && (py != 0)) {
-		mouse.mickeysPerPixel_x = (float)px / X_MICKEY;
-		mouse.mickeysPerPixel_y = (float)py / Y_MICKEY;
-		mouse.pixelPerMickey_x  = X_MICKEY / (float)px;
-		mouse.pixelPerMickey_y  = Y_MICKEY / (float)py;
-	}
-}
-
-static void SetSensitivity(uint16_t px, uint16_t py, uint16_t dspeed)
+    // Try to aggregate move / wheel events
+    if ((event_dos_moved && event.id == MouseEventId::MouseHasMoved) ||
+        (event_dos_wheel && event.id == MouseEventId::WheelHasMoved)) {
+        event.req_dos = false; // DOS queue already has such an event
+        return;
+    }
+
+    // Non-button events can't be aggregated with button events
+    // at this moment, this is only possible once they are being
+    // passed to the interrupt
+    if (event.id == MouseEventId::MouseHasMoved ||
+        event.id == MouseEventId::WheelHasMoved)
+        return;
+
+    // Generate masks to detect whether two button events can be
+    // aggregated (might be needed later even if we have no more
+    // events now)
+    if (event.mask & aggr_mask_pressed)
+        // Set 'pressed+released' for every 'pressed' bit
+        event.aggr_mask = static_cast<uint8_t>(event.mask | (event.mask << 1));
+    else if (event.mask & aggr_mask_released)
+        // Set 'pressed+released' for every 'released' bit
+        event.aggr_mask = static_cast<uint8_t>(event.mask | (event.mask >> 1));
+
+    // If no button events in the queue, skip further processing
+    if (!num_events)
+        return;
+
+    // Try to aggregate with the last queue event
+    auto &last_event = events[static_cast<uint8_t>(idx_first + num_events) % 
+        events.size()];
+    if (GCC_UNLIKELY(last_event.aggr_mask & event.aggr_mask))
+        return; // no aggregation possible
+
+    // Both events can be aggregated into a single one
+    last_event.mask |= event.mask;
+    last_event.aggr_mask |= event.aggr_mask;
+    last_event.buttons_12S = event.buttons_12S;
+
+    // Event aggregated, DOS does not need it any more
+    event.req_dos = false;
+}
+
+MouseEvent &MouseQueue::PopEventBtn()
 {
-	if (px > 100)
-		px = 100;
-	if (py > 100)
-		py = 100;
-	if (dspeed > 100)
-		dspeed = 100;
-	// save values
-	mouse.senv_x_val = px;
-	mouse.senv_y_val = py;
-	mouse.dspeed_val = dspeed;
-	if ((px != 0) && (py != 0)) {
-		px--; // Inspired by cutemouse
-		py--; // Although their cursor update routine is far more
-		      // complex then ours
-		mouse.senv_x = (static_cast<float>(px) * px) / 3600.0f + 1.0f / 3.0f;
-		mouse.senv_y = (static_cast<float>(py) * py) / 3600.0f + 1.0f / 3.0f;
-	}
-}
+    assert(num_events > 0);
 
-static void ResetHardware(void)
-{
-	PIC_SetIRQMask(MOUSE_IRQ, false);
+    auto &event = events[idx_first];
+
+    idx_first++;
```

Comment:
pre-increment/decrement for consistency?

Reply:
Changed.